File: enums.go
package mcm

// Enum values in this file are derived from the OpenAPI specs (v1.1.0)
// downloaded from api.sap.com on 2026-04-13. The real SAP system may
// accept additional codes not listed here. All enums are typed strings,
// so unknown values from the API will still deserialize correctly.

// Division represents the energy division of a measurement concept.
// The division is derived from the referenced measurement concept model and class.
type Division string

const (
	// DivisionElectricity is the electricity division.
	DivisionElectricity Division = "EL"
	// DivisionGas is the gas division.
	DivisionGas Division = "GA"
	// DivisionWater is the water division.
	DivisionWater Division = "WA"
	// DivisionRemoteHeat is the remote heat division.
	DivisionRemoteHeat Division = "RH"
)

// OverallStatus is the combined status of the instance status and
// the process status of a measurement concept instance.
type OverallStatus string

const (
	// OverallStatusInitial indicates the instance has been created but not yet processed.
	OverallStatusInitial OverallStatus = "INITIAL"
	// OverallStatusNew indicates the instance is new and awaiting processing.
	OverallStatusNew OverallStatus = "NEW"
	// OverallStatusError indicates an error occurred during processing.
	OverallStatusError OverallStatus = "ERROR"
	// OverallStatusActive indicates the instance is active and operational.
	OverallStatusActive OverallStatus = "ACTIVE"
	// OverallStatusHistoric indicates the instance has been superseded by a newer version.
	OverallStatusHistoric OverallStatus = "HISTORIC"
	// OverallStatusVersionCancel indicates the instance version has been cancelled.
	OverallStatusVersionCancel OverallStatus = "VERSION_CANCEL"
)

// ClassType represents the type of a measurement concept class.
// SAP templates are read-only reference templates; CLASS is production-usable.
type ClassType string

const (
	// ClassTypeClass is a production-usable measurement concept class.
	ClassTypeClass ClassType = "CLASS"
	// ClassTypeSAPTemplate is a read-only SAP reference template.
	ClassTypeSAPTemplate ClassType = "SAPTEMPLATE"
)

// ConceptType represents the type of a measurement concept model.
// SAP templates are read-only reference templates; MODEL is production-usable.
type ConceptType string

const (
	// ConceptTypeModel is a production-usable measurement concept model.
	ConceptTypeModel ConceptType = "MODEL"
	// ConceptTypeSAPTemplate is a read-only SAP reference template.
	ConceptTypeSAPTemplate ConceptType = "SAPTEMPLATE"
)

// ModelStatus represents the status of a measurement concept model.
type ModelStatus string

const (
	// ModelStatusInProgress indicates the model is being edited.
	ModelStatusInProgress ModelStatus = "IN_PROGRESS"
	// ModelStatusActive indicates the model is active and can be used for instantiation.
	ModelStatusActive ModelStatus = "ACTIVE"
	// ModelStatusDeactivated indicates the model has been deactivated.
	ModelStatusDeactivated ModelStatus = "DEACTIVATED"
)

// Direction represents the energy flow direction of an actor or market location.
// For electricity, both IN (supply/feed-in) and OUT (demand/consumption) are supported.
// For gas, water, and remote heat, only OUT is supported.
type Direction string

const (
	// DirectionIn is the supply / feed-in direction.
	DirectionIn Direction = "IN"
	// DirectionOut is the demand / consumption direction.
	DirectionOut Direction = "OUT"
)

// ActorType represents the type of an actor in a measurement concept.
type ActorType string

const (
	// ActorTypeConsumer is a consumer of energy.
	ActorTypeConsumer ActorType = "CONSUMER"
	// ActorTypeProducer is a producer / generator of energy.
	ActorTypeProducer ActorType = "PRODUCER"
	// ActorTypeStorage is an energy storage unit.
	ActorTypeStorage ActorType = "STORAGE"
)

// MeteringLocationType represents the type of a metering location
// within a measurement concept class.
type MeteringLocationType string

const (
	// MeteringLocationTypeGridMes is a grid measurement point.
	MeteringLocationTypeGridMes MeteringLocationType = "GRIDMES"
	// MeteringLocationTypeSeriesMes is a serial switching measurement point.
	MeteringLocationTypeSeriesMes MeteringLocationType = "SERIESMES"
	// MeteringLocationTypeGeneratorMes is a generator measurement point.
	MeteringLocationTypeGeneratorMes MeteringLocationType = "GENERATORMES"
	// MeteringLocationTypeStorageMes is a storage measurement point.
	MeteringLocationTypeStorageMes MeteringLocationType = "STORAGEMES"
	// MeteringLocationTypeDiffMes is a difference measurement point.
	MeteringLocationTypeDiffMes MeteringLocationType = "DIFFMES"
	// MeteringLocationTypeCompareMes is a comparative measurement point.
	MeteringLocationTypeCompareMes MeteringLocationType = "COMPAREMES"
)

// MeteringTaskType represents the type of a metering task assigned to a metering location.
type MeteringTaskType string

const (
	// MeteringTaskTypeAE is active energy measurement (Arbeitsenergieerfassung).
	MeteringTaskTypeAE MeteringTaskType = "AE"
	// MeteringTaskTypeOV is operating volume measurement (Betriebsvolumen).
	MeteringTaskTypeOV MeteringTaskType = "OV"
	// MeteringTaskTypeEV is energetic value measurement (Energetischer Wert).
	MeteringTaskTypeEV MeteringTaskType = "EV"
)

// MeteringProcedure represents the metering procedure of a metering task.
// Determines how meter readings are collected and processed.
type MeteringProcedure string

const (
	// MeteringProcedureSLP is standard load profile (Standardlastprofil).
	MeteringProcedureSLP MeteringProcedure = "SLP"
	// MeteringProcedureRLM is interval reading (registrierende Leistungsmessung).
	MeteringProcedureRLM MeteringProcedure = "RLM"
	// MeteringProcedureIR is interval recording.
	MeteringProcedureIR MeteringProcedure = "IR"
)

// MarketLocationUsage represents the usage purpose of a market location calculation rule.
type MarketLocationUsage string

const (
	// MarketLocationUsageBilling is used for billing.
	MarketLocationUsageBilling MarketLocationUsage = "BILLING"
	// MarketLocationUsageGridUse is used for grid usage calculation.
	MarketLocationUsageGridUse MarketLocationUsage = "GRIDUSE"
	// MarketLocationUsageOUBill is used for own-use billing.
	MarketLocationUsageOUBill MarketLocationUsage = "OUBILL"
	// MarketLocationUsageREB is used for rebilling.
	MarketLocationUsageREB MarketLocationUsage = "REB"
	// MarketLocationUsageSettle is used for settlement.
	MarketLocationUsageSettle MarketLocationUsage = "SETTLE"
)

// MarketLocationType represents the type of a market location.
type MarketLocationType string

const (
	// MarketLocationTypeSupply is a supply market location.
	MarketLocationTypeSupply MarketLocationType = "SUPPLY"
)

// ProcessType represents the type of a change process for a measurement concept instance.
type ProcessType string

const (
	// ProcessTypeCreate is the initial creation of the instance.
	ProcessTypeCreate ProcessType = "CREATE"
)

// MeteringLocationPurpose represents the purpose of a metering location
// in a measurement concept model.
type MeteringLocationPurpose string

const (
	// MeteringLocationPurposeSC is self-consumption (Eigenverbrauch).
	MeteringLocationPurposeSC MeteringLocationPurpose = "SC"
	// MeteringLocationPurposeCST is commercial settlement transfer.
	MeteringLocationPurposeCST MeteringLocationPurpose = "CST"
)

// ForecastBasis represents the forecast basis of a market location.
type ForecastBasis string

const (
	// ForecastBasisRLM is interval reading based forecast.
	ForecastBasisRLM ForecastBasis = "RLM"
	// ForecastBasisREM is remainder based forecast.
	ForecastBasisREM ForecastBasis = "REM"
	// ForecastBasisHO is household based forecast.
	ForecastBasisHO ForecastBasis = "HO"
)

// MeasuringType represents the measuring type of a metering location in process data.
type MeasuringType string

const (
	// MeasuringTypeCME is conventional metering equipment.
	MeasuringTypeCME MeasuringType = "CME"
	// MeasuringTypeMME is modern metering equipment (smart meter).
	MeasuringTypeMME MeasuringType = "MME"
)

// Rate represents the tariff rate code of a metering task.
type Rate string

const (
	// RateSingleRate is a single rate tariff (Eintarif).
	RateSingleRate Rate = "SR"
	// RateDoubleRate is a double rate tariff (Doppeltarif).
	RateDoubleRate Rate = "DR"
)

File: src/sap_mcm_client/enums.py
"""StrEnum types for all SAP MCM domain codes.

These enum values are derived from the OpenAPI specs (v1.1.0) downloaded from
api.sap.com on 2026-04-13. The real SAP system may accept additional codes
not listed here. All enums use StrEnum so that unknown values from the API
can still be compared as plain strings without breaking deserialization —
Pydantic will accept any string, not just the known members, when the field
type is ``str | SomeEnum`` or when ``extra="ignore"`` is set.
"""

from enum import StrEnum


class Division(StrEnum):
    """The energy division of a measurement concept.

    The supported divisions are electricity, gas, water, and remote heat.
    The division is derived from the referenced measurement concept model and class.
    """

    ELECTRICITY = "EL"
    """Electricity"""
    GAS = "GA"
    """Gas"""
    WATER = "WA"
    """Water"""
    REMOTE_HEAT = "RH"
    """Remote heat"""


class OverallStatus(StrEnum):
    """The combined status of the instance status and the process status of a measurement concept instance."""

    INITIAL = "INITIAL"
    """The instance has been created but not yet processed."""
    NEW = "NEW"
    """The instance is new and awaiting processing."""
    ERROR = "ERROR"
    """An error occurred during processing."""
    ACTIVE = "ACTIVE"
    """The instance is active and operational."""
    HISTORIC = "HISTORIC"
    """The instance has been superseded by a newer version."""
    VERSION_CANCEL = "VERSION_CANCEL"
    """The instance version has been cancelled."""


class ClassType(StrEnum):
    """The type of a measurement concept class.

    SAP templates are read-only reference templates; CLASS is production-usable.
    """

    CLASS = "CLASS"
    """A production-usable measurement concept class."""
    SAP_TEMPLATE = "SAPTEMPLATE"
    """A read-only SAP reference template."""


class ConceptType(StrEnum):
    """The type of a measurement concept model.

    SAP templates are read-only reference templates; MODEL is production-usable.
    """

    MODEL = "MODEL"
    """A production-usable measurement concept model."""
    SAP_TEMPLATE = "SAPTEMPLATE"
    """A read-only SAP reference template."""


class ModelStatus(StrEnum):
    """The status of a measurement concept model."""

    IN_PROGRESS = "IN_PROGRESS"
    """The model is being edited."""
    ACTIVE = "ACTIVE"
    """The model is active and can be used for instantiation."""
    DEACTIVATED = "DEACTIVATED"
    """The model has been deactivated."""


class Direction(StrEnum):
    """The energy flow direction of an actor or market location.

    For the electricity division, both IN (supply/feed-in) and OUT (demand/consumption)
    are supported. For gas, water, and remote heat, only OUT is supported.
    """

    IN = "IN"
    """Supply / feed-in direction."""
    OUT = "OUT"
    """Demand / consumption direction."""


class ActorType(StrEnum):
    """The type of an actor in a measurement concept."""

    CONSUMER = "CONSUMER"
    """A consumer of energy."""
    PRODUCER = "PRODUCER"
    """A producer / generator of energy."""
    STORAGE = "STORAGE"
    """An energy storage unit."""


class MeteringLocationType(StrEnum):
    """The type of a metering location within a measurement concept class.

    Defines the role of the metering location in the circuit plan.
    """

    GRID_MES = "GRIDMES"
    """Grid measurement point."""
    SERIES_MES = "SERIESMES"
    """Serial switching measurement point."""
    GENERATOR_MES = "GENERATORMES"
    """Generator measurement point."""
    STORAGE_MES = "STORAGEMES"
    """Storage measurement point."""
    DIFF_MES = "DIFFMES"
    """Difference measurement point."""
    COMPARE_MES = "COMPAREMES"
    """Comparative measurement point."""


class MeteringTaskType(StrEnum):
    """The type of a metering task assigned to a metering location."""

    AE = "AE"
    """Active energy measurement (Arbeitsenergieerfassung)."""
    OV = "OV"
    """Operating volume measurement (Betriebsvolumen)."""
    EV = "EV"
    """Energetic value measurement (Energetischer Wert)."""


class MeteringProcedure(StrEnum):
    """The metering procedure of a metering task or calculation rule.

    Determines how meter readings are collected and processed.
    """

    SLP = "SLP"
    """Standard load profile (Standardlastprofil)."""
    RLM = "RLM"
    """Interval reading / registrierende Leistungsmessung."""
    IR = "IR"
    """Interval recording."""


class MarketLocationUsage(StrEnum):
    """The usage purpose of a market location calculation rule."""

    BILLING = "BILLING"
    """Used for billing."""
    GRID_USE = "GRIDUSE"
    """Used for grid usage calculation."""
    OU_BILL = "OUBILL"
    """Used for own-use billing."""
    REB = "REB"
    """Used for rebilling."""
    SETTLE = "SETTLE"
    """Used for settlement."""


class MarketLocationType(StrEnum):
    """The type of a market location."""

    SUPPLY = "SUPPLY"
    """A supply market location."""


class ProcessType(StrEnum):
    """The type of a change process for a measurement concept instance."""

    CREATE = "CREATE"
    """Initial creation of the instance."""


class MeteringLocationPurpose(StrEnum):
    """The purpose of a metering location in a measurement concept model."""

    SC = "SC"
    """Self-consumption (Eigenverbrauch)."""
    CST = "CST"
    """Commercial settlement transfer."""


class ForecastBasis(StrEnum):
    """The forecast basis of a market location."""

    RLM = "RLM"
    """Interval reading based forecast."""
    REM = "REM"
    """Remainder based forecast."""
    HO = "HO"
    """Household based forecast."""


class MeasuringType(StrEnum):
    """The measuring type of a metering location in process data."""

    CME = "CME"
    """Conventional metering equipment."""
    MME = "MME"
    """Modern metering equipment (smart meter)."""


class Rate(StrEnum):
    """The tariff rate code of a metering task."""

    SINGLE_RATE = "SR"
    """Single rate tariff (Eintarif)."""
    DOUBLE_RATE = "DR"
    """Double rate tariff (Doppeltarif)."""
